### DISCUSSION TITLE
DeploymentConfig's Healthcheck in regard to waves and phases

### DISCUSSION THREAD

**Discussion by etherealy** [argoproj/argo-cd#10434] [General]
Hello there,

We are in the process of evaluating ArgoCD to industrialize the deployment of our solution, which is constituted of several services, on multiple Openshift clusters (v4.10).  
Most of which are our customers', and whose access to is limited.  

We use the Openhsift GitOps Operator v1.5.5, which installs ArgoCD v2.3.7.  

For our current experiment, we deploy an application containing a database, back, font, and a cache.  
We need to have our cache rolled last in order to be sure that it does not contain obsolete data when we upgrade the application.  
(Or have a script run in post-upgrade to invalidate the cache)

Either way, ArgoCD's waves, and phases seem like a perfect match to handle that kind of thing.

However, when experimenting with it, we noticed that as soon as the ReplicationController is created, the DeploymentConfig is considered healthy by ArgoCD, and therefore the subsequent waves then phases are triggered.

I would have expected the DC to be considered healthy only once the pods' rolling is successful.  
I was surprised by the current behavior in regard to waves and phases.

Is this the intended behavior?  
Am I missing something that explains why this has been implemented as such?
Does changing the behavior so that the DC is only considered healthy once the pods have been rolled (or rather the New ReplicationController is in an Available state)  might degrade other ArgoCD features? 

FYI,
<details>
<summary>The current health.lua for the DC is the following:</summary>

```lua
health_check = {}
if obj.status ~= nil then
  if obj.status.conditions ~= nil and obj.status.replicas ~= nil then
    numTrue = 0
    for i, condition in pairs(obj.status.conditions) do
      if (condition.type == "Available" or condition.type == "Progressing") and condition.status == "True" then
        numTrue = numTrue + 1
      end
    end
    if numTrue == 2 or obj.status.replicas == 0 then
      health_check.status = "Healthy"
      health_check.message = "replication controller successfully rolled out"
      return health_check
    elseif numTrue == 1 then
      health_check.status = "Progressing"
      health_check.message = "replication controller is waiting for pods to run"
      return health_check
    else
      health_check.status = "Degraded"
      health_check.message = "Deployment config is degraded"
      return health_check
    end
  end
end
health_check.status = "Progressing"
health_check.message = "replication controller is waiting for pods to run"
return health_check
```

</details>

<details>
<summary>I achieved the desired behavior by updating the health.lua as such:</summary>

```diff
health_check = {}
if obj.status ~= nil then
  if obj.status.conditions ~= nil and obj.status.replicas ~= nil then
    numTrue = 0
    for i, condition in pairs(obj.status.conditions) do
-      if (condition.type == "Available" or condition.type == "Progressing") and condition.status == "True" then
+      if (condition.type == "Available" or (condition.type == "Progressing" and condition.reason == "NewReplicationControllerAvailable")) and condition.status == "True" then
        numTrue = numTrue + 1
      end
    end
    if numTrue == 2 or obj.status.replicas == 0 then
      health_check.status = "Healthy"
      health_check.message = "replication controller successfully rolled out"
      return health_check
    elseif numTrue == 1 then
      health_check.status = "Progressing"
      health_check.message = "replication controller is waiting for pods to run"
      return health_check
    else
      health_check.status = "Degraded"
      health_check.message = "Deployment config is degraded"
      return health_check
    end
  end
end
health_check.status = "Progressing"
health_check.message = "replication controller is waiting for pods to run"
return health_check
```

</details>

Does the modification to the DC's health.lua seems appropriate? 

( If there are any RedHatters in the place :) )

Thank you for any pointers or thoughts you might have on this issue :)

**Comment by iam-veeramalla** [comment]
Hi @etherealy , I totally agree with this. DeploymentConfig should be considered healthy either when the `deployer pod is in completed state` or `replicationController is in available state`.

If you are ok, Please go ahead and raise a PR for this or I don't mind raising a PR on your behalf :). Thanks.

**Reply by etherealy** [reply]
Hello @iam-veeramalla, thanks for your answer !

I'll raise a PR with your blessing then ;)
